server.rb

```diff
@@ -0,0 +1,189 @@
+        xml.root do
+          xml.message do
+            xml.status status.flatten.first
+          end
+        end
+      end
+      [200, builder.to_xml]
+    else
+      [400]
+    end
+  end
+
+  def db
+    SQLite3::Database.new "test.db"
+  end
+
+  class JSONMessage
+
+    def initialize(data)
+      @to, @from, @message = parse(data)
+    end
+
+    def parse(data)
+      parsed_json = JSON.parse(data)
+      [parsed_json['to'], parsed_json['from'], parsed_json['message']]
+    end
+
+    def valid?
+      return false if [@to, @from, @message].any?(&:nil?)
+      return false if db.execute("SELECT * from users where id = '#{@to}' LIMIT 1").empty?
+      return false if db.execute("SELECT * from users where id = '#{@from}' LIMIT 1").empty?
+      true
+    end
+
+    def save
+      @id = SecureRandom.uuid
+      db.execute(<<~SQL
+        INSERT INTO messages(id, to_user, from_user, message, status)
+        VALUES('#{@id}', '#{@to}','#{@from}', '#{@message}', 'pending');
+        SQL
+      )
+      @id
+    end
+
+    def send_message
+      sent = ExternalApiClient.send_message({to: @to, from: @from, message: @message})
+      if sent
+        db.execute(<<-SQL
+          UPDATE messages
+          SET status = 'sent'
+          WHERE id = '#{@id}';
+        SQL
+                  )
+      else
+        db.execute(<<-SQL
+          UPDATE messages
+          SET status = 'failed'
+          WHERE id = '#{@id}';
+        SQL
+                  )
+      end
+    end
+
+    def db
+      SQLite3::Database.new "test.db"
+    end
+  end
+
+  class XMLMessage
+
+    def initialize(data)
+      @to, @from, @message = parse(data)
+    end
+
+    def parse(data)
+      parsed_xml = Nokogiri::XML(data)
+      to = parsed_xml.xpath('//to').text
+      from = parsed_xml.xpath('//from').text
+      message = parsed_xml.xpath('//message').text
+      [to, from, message]
+    end
+
+    def valid?
+      return false if [@to, @from, @message].any?(&:nil?)
+      return false if db.execute("SELECT * from users where id = '#{@to}' LIMIT 1").empty?
+      return false if db.execute("SELECT * from users where id = '#{@from}' LIMIT 1").empty?
+      true
+    end
+
+    def save
+      @id = SecureRandom.uuid
+      db.execute(<<~SQL
+        INSERT INTO messages(id, to_user, from_user, message, status)
+        VALUES('#{@id}', '#{@to}','#{@from}', '#{@message}', 'pending');
+        SQL
+      )
+      @id
+    end
+
+    def send_message
+      sent = ExternalApiClient.send_message({to: @to, from: @from, message: @message})
+      if sent
+        db.execute(<<-SQL
+          UPDATE messages
+          SET status = 'sent'
+          WHERE id = '#{@id}';
+        SQL
+                  )
+      else
+        db.execute(<<-SQL
+          UPDATE messages
+          SET status = 'failed'
+          WHERE id = '#{@id}';
+        SQL
+                  )
+      end
+    end
+
+    def db
+      SQLite3::Database.new "test.db"
+    end
+  end
```

Comment:
`JSONMessage` and `XMLMessage` classes have a lot of duplicate code. The only difference is how the input is parsed. I would suggest having `Message` class responsible for validating, saving, and sending the message and using a separate class like `MessageParser` to parse the message based on the request/response format. See the comment regarding the how the controllers are written for this to make more sense.